File: arrays_and_hashing/product_of_array.rb
# frozen_string_literal: true

# Time complexity: O(n)
# Space complexity: O(n)
def product_except_self(nums)
  length = nums.length
  pre = Array.new(length)
  pre[0] = 1
  post = Array.new(length)
  post[-1] = 1
  prod = Array.new(length)

  (1..length - 1).each do |i|
    pre[i] = pre[i - 1] * nums[i - 1]
    post[length - 1 - i] = post[length - i] * nums[length - i]
  end

  (0..length - 1).each do |i|
    prod[i] = pre[i] * post[i]
  end

  prod
end

# Solution with lower memory
def product_except_self(nums)
  length = nums.length
  prod = Array.new(length)

  0.upto(length - 1).each do |i|
    prod[i] = if i == 0
                prod[i] = 1
              else
                prod[i] = prod[i - 1] * nums[i - 1]
              end
  end

  post = 1
  (length - 1).downto(0).each do |i|
    prod[i] *= post
    post *= nums[i]
  end
  
  prod
end

File: arrays_and_hashing/product_of_array.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)

        pre = [1] * length
        for i in range(1, length):
            pre[i] = nums[i - 1] * pre[i - 1]

        suff = [1] * length
        for i in range(length - 1, 0, -1):
            suff[i - 1] = nums[i] * suff[i]

        res = []
        for i in range(length):
            res.append(pre[i] * suff[i])
        
        return res